lib/part.ml

```diff
@@ -85,33 +74,42 @@ module Parse_parts = struct
   let rec parse_parts input make_part current_part part_lines nline =
     let open Util.Result.Infix in
     let nline = nline + 1 in
-    let line = input_line_err input in
-    match (parse_line line, current_part) with
-    | Normal line, _ ->
-        parse_parts input make_part current_part (line :: part_lines) nline
-    | Part_end, Some _ ->
-        parse_parts input anonymous_part None [] nline
-        >>| List.cons (make_part ~is_begin_end_part:true part_lines)
-    | Part_end, None -> Error ("There is no part to end.", nline)
-    | Part_begin (next_part_name, sep_indent), None ->
-        let next_part = next_part ~name:next_part_name ~sep_indent in
-        let rcall =
-          parse_parts input next_part (Some next_part_name) [] nline
-        in
-        if part_lines = [] then rcall
-          (* Ignore empty anonymous parts: needed for legacy support *)
-        else rcall >>| List.cons (make_part ~is_begin_end_part:true part_lines)
-    | Compat_attr (name, sep_indent), None ->
-        let next_part = next_part ~name ~sep_indent in
-        parse_parts input next_part None [] nline
-        >>| List.cons (make_part ~is_begin_end_part:false part_lines)
-    | Part_begin _, Some p | Compat_attr _, Some p ->
-        let msg = Printf.sprintf "Part %s has no end." p in
-        Error (msg, nline)
-    | File_end, Some p ->
-        let msg = Printf.sprintf "File ended before part %s." p in
-        Error (msg, nline)
-    | File_end, None -> Ok [ make_part ~is_begin_end_part:true part_lines ]
+    match input_line_err input with
```

Comment:
I think I find the new implementation, with nested pattern matching, a bit harder to follow.

Maybe we could extract a separate, mutually recursive function to handle the `Ok _` case.

What do you think?

Reply:
I find it a bit easier to follow because it really had two concerns:

1. EOF handling
2. Parsing

`File_end` was not a result of parsing, it is just a signal to the parser to stop. But admittedly, the function is quite complex and nested (and it took me a while to understand), so you correctly point out that that it still conflates parsing and IO. It's a good point, let me try to improve it.

Reply:
I rewrote `parse_parts` further:

1. Instead of reading from the channel, I read into a list, thus the pattern match on the `End_of_file` marker became a pattern match on `[]`, that is, the end of the input. This of course means it still has to deal with these two cases.
2. Then I realized that the function doesn't need to be recursive. What was making the function hard to read for me was the recursive calls with `>>|` to cons things to the list. I noted that `fold_left` can do the same thing:
    * Supply an `element` to work on and terminate when done
    * Deal with the state in the accumulator
Thus thus the function could be split into "deal with the inputs" via `fold_left` and "deal with the tail" based on the result of the accumulator. Given our minimum version is 4.08, `let*` helps to unwrap the arguments.
3. Finally, given I changed from a version that lazily reads out of an `in_channel` to a version that eagerly reads the input into a `list`, I thought I can put the laziness back in by switching from `List.fold_left` to `Seq.fold_left` which was indeed a drop-in replacement.

I believe the resulting function is a bit easier to understand given each piece of code only deals with one aspect and the fold is a bit simpler to follow through than the recursion. Fortunately the tests are pretty comprehensive so I immediately got feedback when some part of the logic was implemented incorrectly.